### PR TITLE
Import API to APIM using API Operator

### PR DESCRIPTION
**Pull request by RakhithaRR**
### Purpose
This PR introduces importing an API to APIM using the Operator. This can be done either by providing a configmap containing a zip file of an API project or by providing a configmap containing the swagger definition of an API.

To import the API, user should set the value `importAPIToAPIManager: "true"` in `controller-config` and set the relevant REST API configs in `apim-config` for the APIM instance.

```
  #The following 3 endpoints are related to importing API to API Manager
  #API Manager Keymanager/DCR endpoint
  apimKeymanagerEndpoint: "https://localhost:9443"
  #API Manager Publisher endpoint
  apimPublisherEndpoint: "https://localhost:9443"
  #API Manager token endpoint
  apimTokenEndpoint: "https://localhost:8243/token"
  #Skip verification for the REST API invocations. If "false", you need to provide the cert
  insecureSkipVerify: "true"
  #Secret name containing APIM credentials and cert
  apimCredentialsSecret: "apim-secret"
```

Related issue: https://github.com/wso2/k8s-api-operator/issues/481
Suggested label: Type/New Feature

### CODE REVIEW

File: api-operator/pkg/apim/rest.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package apim
```
Comment (pubudu538) [inline api-operator/pkg/apim/rest.go:1]: Add license header.

File: api-operator/pkg/utils/file.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package utils
```
Comment (pubudu538) [inline api-operator/pkg/utils/file.go:1]: Add license header

**Review by pubudu538 — CHANGES_REQUESTED** [review]

File: api-operator/pkg/apim/constants.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+	adminAPIImportEndpoint					= "api/am/admin/v1/import/api"
+	defaultClientRegistrationEndpointSuffix = "client-registration/v0.17/register"
+	defaultApiListEndpointSuffix 			= "api/am/publisher/v1/apis"
+	defaultTokenEndpoint					= "oauth2/token"
+	importAPIFromSwaggerEndpoint			= "api/am/publisher/v1/apis/import-openapi"
+)
+
+// APIDefinition represents an API artifact in APIM
+type APIDefinition struct {
```
[the remark this answers is not in the document]
Comment (pubudu538) [reply api-operator/pkg/apim/constants.go:51]: Plz ignore this comment.

File: api-operator/pkg/apim/deleteAPI.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	apiId, err := getAPIId(token, endpoint+"/"+defaultApiListEndpointSuffix, name, version)
+	if err != nil {
+		return err
+	}
+
+	deleteErr := deleteAPIById(endpoint, apiId, token)
+	if deleteErr != nil {
+		logDelete.Error(deleteErr, "Error when deleting the API from APIM")
```
Comment (renuka-fernando) [inline api-operator/pkg/apim/deleteAPI.go:131]: Shouldn't this error be passed to the caller

File: api-operator/pkg/controller/api/api_controller.go [hunk 5]
```diff
@@ -670,6 +672,18 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
 		r.recorder.Event(instance, corev1.EventTypeNormal, "Deploy",
 			fmt.Sprintf("Successfully deployed the API: %s.", instance.Name))
 
+		if importAPIEnabled {
+			importErr := apim.ImportAPI(&r.client, instance)
+			if importErr != nil {
+				r.recorder.Event(instance, eventTypeError, "Import",
```
Comment (renuka-fernando) [inline api-operator/pkg/controller/api/api_controller.go:678]: Shall we rename reason for event, something like
```suggestion
				r.recorder.Event(instance, eventTypeError, "FailedAPIMImport",
```

ref: https://www.bluematador.com/blog/kubernetes-events-explained

File: api-operator/pkg/controller/api/api_controller.go [hunk 6]
```diff
@@ -670,6 +672,18 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+					fmt.Sprintf("Error occured while importing the API to APIM"))
+				return reconcile.Result{}, importErr
+			}
+			r.recorder.Event(instance, corev1.EventTypeNormal, "Import",
```
Comment (renuka-fernando) [inline api-operator/pkg/controller/api/api_controller.go:682]: Shall we rename reason for this event also, something like "APIMImported" or "ImportedToAPIM"

File: api-operator/pkg/controller/api/api_controller.go [hunk 7]
```diff
@@ -209,6 +210,7 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
 	mgwDockerImage.RegistryType = registry.Type(dockerRegistryConf.Data[registryTypeConst])
 	mgwDockerImage.RepositoryName = dockerRegistryConf.Data[repositoryNameConst]
 	operatorMode := controlConfigData[operatorModeConst]
+	importAPIEnabled, err := strconv.ParseBool(controlConfigData[importAPIEnabledConst])
```
Comment (renuka-fernando) [inline api-operator/pkg/controller/api/api_controller.go:213]: Shall we handle this error `err`

Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go:213]: Fixed in e0400b3

File: api-operator/pkg/controller/api/api_controller.go (hunk 6 above)
Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go]: Fixed in e0400b3

File: api-operator/pkg/controller/api/api_controller.go (hunk 5 above)
Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go]: Fixed in e0400b3

File: api-operator/pkg/apim/deleteAPI.go (hunk 4 above)
Comment (RakhithaRR) [reply api-operator/pkg/apim/deleteAPI.go:147]: Fixed in e0400b3

File: api-operator/pkg/utils/file.go (hunk 2 above)
Comment (RakhithaRR) [reply api-operator/pkg/utils/file.go:17]: Fixed in e0400b3

File: api-operator/pkg/apim/rest.go (hunk 1 above)
Comment (RakhithaRR) [reply api-operator/pkg/apim/rest.go:17]: Fixed in e0400b3

**Review by pubudu538 — APPROVED** [review]

File: api-operator/pkg/apim/deleteAPI.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+		return deleteErr
+	}
+
+	return nil
+}
```
Comment (dinusha92) [inline api-operator/pkg/apim/deleteAPI.go:152]: Shall we add a new line at the end?

File: api-operator/deploy/controller-configs/controller_conf.yaml [hunk 9]
```diff
@@ -57,6 +57,8 @@ data:
   operatorMode: "default"
   #Expose custom metrics. Default-> observabilityEnabled: "false"
   observabilityEnabled: "false"
+  #Import an API to APIM through Operator
+  importAPIEnabled: "false"
```
[the remark this answers is not in the document]
Comment (RakhithaRR) [reply api-operator/deploy/controller-configs/controller_conf.yaml]: Fixed

File: api-operator/deploy/controller-configs/controller_conf.yaml [hunk 10]
```diff
@@ -84,6 +86,8 @@ data:
   #Ports from which the managed API service is getting exposed
   httpPort: "9090"
   httpsPort: "9095"
+  #API Manager url for importing APIs using Operator
+  apimEndpoint: "https://localhost:9443"
```
[the remark this answers is not in the document]
Comment (RakhithaRR) [reply api-operator/deploy/controller-configs/controller_conf.yaml]: Fixed

File: api-operator/pkg/apim/deleteAPI.go (hunk 8 above)
Comment (RakhithaRR) [reply api-operator/pkg/apim/deleteAPI.go:152]: Fixed in latest code

**Pull request merged**